### PR TITLE
rgw: modify string match_wildcards with fnmatch

### PR DESCRIPTION
**Pull request by qiuxinyidian**
Fixes: https://tracker.ceph.com/issues/62292

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by qiuxinyidian** [comment]
arm make env  issue

**Comment by qiuxinyidian** [comment]
anybody ?

File: src/rgw/rgw_string.cc [hunk 1]
```diff
@@ -2,44 +2,20 @@
...
-      return false;
-    if (*it1 == '?' || eq(*it1, *it2)) {
-      ++it1;
-      ++it2;
-      continue;
-    }
+  if (case_insensive) {
+    flag = FNM_CASEFOLD;
+  }
+  if (fnmatch(pattern.data(), input.data(), flag) == 0) {
```
Comment (cbodley) [inline src/rgw/rgw_string.cc:16]: unfortunately, we can't assume that these string_views are null-terminated

**Review by cbodley — COMMENTED** [review]
according to the examples in https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_resource.html#reference_policies_elements_resource_wildcards, this matching should be greedy

File: src/rgw/rgw_string.cc (hunk 1 above)
Comment (qiuxinyidian) [reply src/rgw/rgw_string.cc:16]: here  we change  string_views to string  maybe OK ？

Comment (qiuxinyidian) [reply src/rgw/rgw_string.cc:16]: > unfortunately, we can't assume that these string_views are null-terminated

in http://web.mit.edu/freebsd/csup/fnmatch.c see

#define	EOS	'\0'

static const char *rangematch(const char *, char, int);

int
fnmatch(const char *pattern, const char *string, int flags)
{
	const char *stringstart;
	char c, test;

	for (stringstart = string;;)
		switch (c = *pattern++) {
		case EOS:
			if ((flags & FNM_LEADING_DIR) && *string == '/')
				return (0);
			return (*string == EOS ? 0 : FNM_NOMATCH);
		case '?':

Comment (mattbenjamin) [reply src/rgw/rgw_string.cc:16]: well for me 1) maybe; 2) probably not?

Comment (qiuxinyidian) [reply src/rgw/rgw_string.cc:16]: is fnmatch() is a good fit for this, or should we extend the existing matching logic? @mattbenjamin @adamemerson

Comment (qiuxinyidian) [reply src/rgw/rgw_string.cc:16]: change it to string ?

Comment (qiuxinyidian) [reply src/rgw/rgw_string.cc:16]: change it to const std::string in match_wildcards()?

Comment (adamemerson) [reply src/rgw/rgw_string.cc:16]: This should be fine. I wanted to check that there wasn't a good way to make Boost Regex greedier, but there doesn't seem to be.

Comment (adamemerson) [reply src/rgw/rgw_string.cc:16]: Though I think I'd prefer extending the existing logic to extra allocations.

Comment (qiuxinyidian) [reply src/rgw/rgw_string.cc:16]: > Though I think I'd prefer extending the existing logic to extra allocations.

@adamemerson @mattbenjamin  who will extend the existing logic to extra allocations ?

Comment (qiuxinyidian) [reply src/rgw/rgw_string.cc:16]: @cbodley  @adamemerson  @mattbenjamin  
What should we do next ?
Maybe one of us  should extend the existing logic to extra allocations.

Comment (cbodley) [reply src/rgw/rgw_string.cc:16]: @qiuxinyidian sorry for the delays. i do think we should match aws' behavior for this parsing, and i'd rather not make this parsing too complicated to in the process. so i do like your approach to use fnmatch()

> if we are going to switch to `fnmatch()`, then the arguments to `match_wildcards()` and `match_policy()` should change to `const char*`s or `const std::string&`s to reflect that requirement

can you try this way?

Comment (qiuxinyidian) [reply src/rgw/rgw_string.cc:16]: @cbodley OK, i will try

Comment (qiuxinyidian) [reply src/rgw/rgw_string.cc:17]: @cbodley  src/rgw/rgw_iam_policy.cc has little confilict ,can you help me fix it?

**Comment by qiuxinyidian** [comment]
@cbodley please check it .

**Comment by qiuxinyidian** [comment]
@adamemerson please check it

**Review by cbodley — APPROVED** [review]

**Comment by cbodley** [comment]
passed qa in https://pulpito.ceph.com/cbodley-2024-01-09_00:24:23-rgw-wip-cbodley-testing-distro-default-smithi/

**Comment by cbodley** [comment]
thanks @qiuxinyidian!

**Pull request merged**